src/client/connstring.rs

```diff
@@ -0,0 +1,318 @@
+use std::ascii::AsciiExt;
+use std::collections::HashMap;
+use std::vec::Vec;
+
+pub const DEFAULT_PORT: u16 = 27017;
+pub const URI_SCHEME: &'static str = "mongodb://";
+
+/// MongoDB connection types.
+pub enum ConnectionType {
+    Invalid,
+    Master,
+    Pair,
+    Set,
+}
+
+/// Encapsulates the hostname and port of a host.
+pub struct Host {
+    pub host_name: String,
+    pub port: u16,
+}
+
+impl Host {
+    /// Creates a new Host struct.
+    pub fn new(host_name: String, port: u16) -> Host {
+        Host {
+            host_name: host_name,
+            port: port,
+        }
+    }
+}
+
+/// Encapsulates the options and read preference tags of a MongoDB connection.
+pub struct ConnectionOptions {
+    pub options: HashMap<String, String>,
+    pub read_pref_tags: Vec<String>,
+}
+
+impl ConnectionOptions {
+    /// Creates a new ConnectionOptions struct.
+    pub fn new(options: HashMap<String, String>, read_pref_tags: Vec<String>) -> ConnectionOptions {
+        ConnectionOptions {
+            options: options,
+            read_pref_tags: read_pref_tags,
+        }
+    }
+
+    /// Retrieves an option using a borrowed String
+    pub fn get(&self, key: &String) -> Option<&String> {
+        self.options.get(key)
+    }
+
+    /// Retrieves an option using a borrowed str
+    pub fn get_str(&self, key: &str) -> Option<&str> {
+        match self.options.get(&key.to_owned()) {
+            Some(val) => Some(&val),
+            None => None,
+        }
+    }
+}
+
+/// Encapsulates information for connection to a single MongoDB host or replicated set.
+pub struct ConnectionString {
+    ctype: ConnectionType,
```

Comment:
I'd recommend just removing it, we don't really use it internally, and it doesn't map well to how we want to think about connection strings going forward.